src/features/abi-methods/data/index.ts

```diff
@@ -0,0 +1,104 @@
+import { ApplicationId } from '@/features/applications/data/types'
+import { jsonAsArc32AppSpec } from '@/features/abi-methods/mappers'
+import { useCallback } from 'react'
+import { useAtomCallback } from 'jotai/utils'
+import { dbConnection } from '@/features/common/data/indexed-db'
+import { AppSpecVersion } from '@/features/abi-methods/data/types'
+import { invariant } from '@/utils/invariant'
+import { writableAtomCache } from '@/features/common/data'
+import { atom } from 'jotai'
+
+const getAppSpecs = async (applicationId: ApplicationId) => {
+  invariant(dbConnection, 'dbConnection is not initialised')
+  return (await (await dbConnection).get('applications-app-specs', applicationId.toString())) ?? []
+}
+const writeAppSpecs = async (applicationId: ApplicationId, appSpecs: AppSpecVersion[]) => {
+  invariant(dbConnection, 'dbConnection is not initialised')
+  if (!appSpecs) return
+
+  await (await dbConnection).put('applications-app-specs', appSpecs, applicationId.toString())
+}
+
+const writableAppSpecVersionsAtom = (applicationId: ApplicationId) => {
+  const appSpecVersions = atom<AppSpecVersion[]>([])
+
+  return atom(
+    (get) => {
+      const value = get(appSpecVersions)
+      if (!value.length) {
+        return getAppSpecs(applicationId)
+      }
+      return value
+    },
+    async (_, set, appSpecs: AppSpecVersion[]) => {
+      await writeAppSpecs(applicationId, appSpecs)
+      set(appSpecVersions, appSpecs)
+    }
+  )
```

Comment:
This will result in use going to indexed db each time when we don't have any app specs.
We only need to do that on the first call.

To solve this we could have an `undefined` state for `appSpecVersions` and use an effect to set that atom to empty when we read and no results come back from the db and the previous value is undefined.